sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala

```diff
@@ -174,10 +174,19 @@ private[sql] object JDBCRelation extends Logging {
     (dialect.quoteIdentifier(column.name), column.dataType)
   }
 
-  private def toInternalBoundValue(value: String, columnType: DataType): Long = columnType match {
-    case _: NumericType => value.toLong
-    case DateType => DateTimeUtils.fromJavaDate(Date.valueOf(value)).toLong
-    case TimestampType => DateTimeUtils.fromJavaTimestamp(Timestamp.valueOf(value))
+  private def toInternalBoundValue(value: String, columnType: DataType): Long = {
+    def parse[T](f: UTF8String => Option[T]): T = {
+      f(UTF8String.fromString(value)).getOrElse {
+        throw new IllegalArgumentException(
+          s"Cannot parse the bound value $value as ${columnType.catalogString}")
+      }
+    }
+    columnType match {
+      case _: NumericType => value.toLong
+      case DateType => parse(stringToDate).toLong
+      case TimestampType =>
+        parse(stringToTimestamp(_, getTimeZone(SQLConf.get.sessionLocalTimeZone)))
```

Comment:
We decide to adjust the timestamp constants based on the user-specified local timezone?

Reply:
cc @cloud-fan

Reply:
Previously we call `Timestamp.valueOf(value)`, which uses JVM local timezone. It seems to me that using Spark session timezone is better.

Reply:
Actually we have to do it. This is a followup of https://github.com/apache/spark/pull/23391 , which changed how we turn the timestamp boundaries to string. Here we change hoow we turn string to timestamp.

Reply:
This is a behavior change. We need to clearly document which inputs start respecting our Spark local session timezone?

Reply:
How about we mention something like .. all timestamps will respect Session timezone, JDBC lower/upper bounds, blabla, ..., and java 8 time will be consistently used across code base .. after the sub-tasks in the umbrella are resolved?